active/0000-keyword-arguments.md

```diff
@@ -0,0 +1,139 @@
+In languages like PHP, default parameters must come at the end of the function and must have a value.
+
+You can have optional parameters in the beginning, middle and end of functions without any default value.
+This means that this design allows for truly optional parameters without any run-time cost of using an `Option`.
+
+# Detailed design
+Currently Rust has
+
+```rust
+fn slice(&self, begin: usize, end: usize) -> &'a str
+fn slice_from(&self, begin: usize) -> &'a str
+fn slice_to(&self, end: usize) -> &'a str
+```
+
+This can be changed to
+
+```rust
+fn slice(&self, from => begin: usize, to => end: usize) -> &'a str
+fn slice(&self, from => begin: usize) -> &'a str
+fn slice(&self, to => end: usize) -> &'a str
+```
+
+Note that these are three different functions that have three different signatures.
+The keywords a function accepts is part of its signature. You can call these functions like this:
+
+```rust
+foo.slice(from => 5); //equivalent to current foo.slice_from(5)
+foo.slice(to => 9);   //equivalent to current foo.slice_to(9)
+foo.slice(from => 5, to => 9);       //equivalent to current foo.slice(5, 9)
+foo.slice(from => 9, to => 5);       //equivalent to current foo.slice(5, 9)
+```
+
+The trait looks like
+
+```rust
+Fn(&self, from => begin: usize, to => end: usize) -> &'a str
+//which desugars to
+Fn<({from: usize, to: usize}, &Self), &'a str>
+```
+
+So this is equivalent for using a `struct` to pass around your keyword parameters.
+But now you are able to use the same function name several times as long as the keywords are different.
+
+# Drawbacks
+
+This is a more complicated design than just having default arguments and overloading.
+Now there are both positional and keyword arguments and they interact with lifetimes, traits, and closures.
+
+# Alternatives
+
+A better design to the above function might be designing it like so:
+
+```rust
+let title = "Title";
+let position = Position(20, 50);
+let dimensions = Dimension(100, 50);
+window.addNewControlDrawingNow(title, position, dimensions);
+```
+
+Now the function takes three parameters, and we assigned them meaningful names.
+Instead of passing a boolean, we created functions that do different things.
+They could be named something like `addNewControl` and `addNewControlDrawingNow`.
+
+While this design is better, it still doesn't solve the problem of having to remember of the order.
+Where do you put `dimensions`, `position`, and the `title`?
+At least the compiler will now verify that the types are correct.
+It is still up to the programmer to actually name those variables well, instead of the API specifying what the keywords should be.
+
+If keyword arguments themselves are not implemented, then there's also the issue of overloading to enable better API design.
+Another proposal would be to assign keywords to all current Rust code with their local names.
```

Comment:
This really threw me for a loop: "assign keywords".  This implies that you want to make all existing argument identifiers reserved keywords in the language or something.  Unfortunate overloading of the term, I suppose.

Maybe reword to something like: "allow any argument to optionally be used as a keyword argument".